### PR TITLE
Add validation for SPV_NV_cooperative_matrix

### PR DESCRIPTION
**Pull request by jeffbolznv**
WIP because this will fail CI until https://github.com/KhronosGroup/SPIRV-Headers/pull/94 is merged and DEPS is updated to use it.

### CODE REVIEW

**Comment by dneto0** [comment]
Bots will all fail until https://github.com/KhronosGroup/SPIRV-Headers/pull/94 is merged.

There is also a minor code formatting violation.  If needed we can take care of that in a cloned PR.

**Comment by jeffbolznv** [comment]
I'll fix the formatting issue when I update the spirv-headers hash. I must have just missed it when I ran clang-format. To my surprise, if I run clang-format on the whole tree on my system, it gives me lots of unexpected diffs...

**Comment by dneto0** [comment]
Compiling with the headers update yields a few nuisance warnings that we force as errors:

/usr/bin/c++  -DSPIRV_CHECK_CONTEXT -DSPIRV_COLOR_TERMINAL -DSPIRV_LINUX -DSPIRV_TIMER_ENABLED -Ithird_party/spirv-tools -Ithird_party/spirv-tools/include -Ithird_party/spirv-tools -Ithird_party/spirv-headers/include -O3 -DNDEBUG -fPIC   -Wall -Wextra -Wnon-virtual-dtor -Wno-missing-field-initializers -Werror -std=c++11 -fno-exceptions -fno-rtti -Wno-long-long -Wshadow -Wundef -Wconversion -Wno-sign-conversion -std=gnu++11 -MD -MT third_party/spirv-tools/source/CMakeFiles/SPIRV-Tools.dir/val/validate_memory.cpp.o -MF third_party/spirv-tools/source/CMakeFiles/SPIRV-Tools.dir/val/validate_memory.cpp.o.d -o third_party/spirv-tools/source/CMakeFiles/SPIRV-Tools.dir/val/validate_memory.cpp.o -c third_party/spirv-tools/source/val/validate_memory.cpp
third_party/spirv-tools/source/val/validate_memory.cpp: In function ‘spv_result_t spvtools::val::{anonymous}::ValidateCooperativeMatrixLoadStoreNV(spvtools::val::ValidationState_t&, const spvtools::val::Instruction*)’:
third_party/spirv-tools/source/val/validate_memory.cpp:1325:31: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
   if (inst->operands().size() > memory_access_index) {
       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
third_party/spirv-tools/source/val/validate_memory.cpp:1251:14: error: unused variable ‘component_type_id’ [-Werror=unused-variable]
   const auto component_type_id = matrix_type->GetOperandAs<uint32_t>(1);
              ^~~~~~~~~~~~~~~~~ 
third_party/spirv-tools/source/val/validate_memory.cpp: At global scope:
third_party/spirv-tools/source/val/validate_memory.cpp:1195:14: error: ‘spv_result_t spvtools::val::{anonymous}::ValidateCooperativeMatrixLengthNV(spvtools::val::ValidationState_t&, const spvtools::val::Instruction*)’ defined but not used [-Werror=unused-function]
 spv_result_t ValidateCooperativeMatrixLengthNV(ValidationState_t& state,
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
third_party/spirv-tools/source/val/validate_memory.cpp: In function ‘spv_result_t spvtools::val::{anonymous}::ValidateCooperativeMatrixLoadStoreNV(spvtools::val::ValidationState_t&, const spvtools::val::Instruction*)’:
third_party/spirv-tools/source/val/validate_memory.cpp:1237:39: error: ‘type_id’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
   auto matrix_type = _.FindDef(type_id);
                                       ^

File: source/val/validate_memory.cpp [hunk 1]
```diff
@@ -1136,6 +1192,144 @@ spv_result_t ValidateArrayLength(ValidationState_t& state,
   return SPV_SUCCESS;
 }
 
+spv_result_t ValidateCooperativeMatrixLengthNV(ValidationState_t& state,
```
Comment (dneto0) [inline source/val/validate_memory.cpp]: Compiler says this function is unused.  Indicates a missing check and missing test coverage.

Comment (jeffbolznv) [reply source/val/validate_memory.cpp]: I've hooked it up and added three tests (bad result type, bad operand type, good). I've also (hopefully) fixed the warnings.

**Comment by jeffbolznv** [comment]
This is passing CI now and ready to merge.

**Comment by jeffbolznv** [comment]
Can this be merged today?

**Pull request merged**